File: holder/src/main/java/com/dhy/apiholder/ApiHolderUtil.java
package com.dhy.apiholder;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;
import java.util.HashMap;
import java.util.Map;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.ObservableTransformer;
import io.reactivex.Single;
import io.reactivex.SingleSource;
import io.reactivex.SingleTransformer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public class ApiHolderUtil {
    private Retrofit retrofit;
    private boolean isAndroidApi;
    private final Map<Class, Object> apis = new HashMap<>();
    private final Map<Class, String> baseUrls = new HashMap<>();
    private AutoCmdUtil autoCmdUtil;

    protected boolean isAndroidApi() {
        return true;
    }

    protected boolean validateEagerly() {
        return false;
    }

    protected OkHttpClient getClient() {
        return new OkHttpClient();
    }

    protected Retrofit getRetrofit(@NonNull OkHttpClient client) {
        return new Retrofit.Builder()
                .validateEagerly(validateEagerly())
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .baseUrl("http://www.demo.com/")
                .client(client)
                .build();
    }

    private void init() {
        this.isAndroidApi = isAndroidApi();
        this.retrofit = getRetrofit(getClient());
        try {
            autoCmdUtil = new AutoCmdUtil();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public final <HOLDER> HOLDER createHolderApi(@NonNull Class<HOLDER> apiHolder) {
        init();
        Class[] partApis = apiHolder.getInterfaces();
        for (Class api : partApis) {
            updateApi(api);
        }
        return createHolderApi(apis, apiHolder);
    }

    /**
     * update api with @{@link BaseUrl}
     */
    public final <API> void updateApi(@NonNull Class<API> api) {
        updateApi(api, null);
    }

    public final <API> void updateApi(@NonNull Class<API> apiClass, @Nullable String baseUrl) {
        Retrofit.Builder retrofitBuilder = retrofit.newBuilder();
        initTimeout(retrofitBuilder, apiClass);
        baseUrl = getBaseUrl(apiClass, baseUrl);
        Retrofit retrofit = retrofitBuilder.baseUrl(baseUrl).build();
        apis.put(apiClass, retrofit.create(apiClass));
        baseUrls.put(apiClass, baseUrl);

        if (validateEagerly() && autoCmdUtil != null) {
            try {
                autoCmdUtil.initHeaders(retrofit);
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            }
        }
    }

    private <API> void initTimeout(Retrofit.Builder retrofitBuilder, @NonNull Class<API> apiClass) {
        if (apiClass.isAnnotationPresent(Timeout.class)) {
            Timeout timeout = apiClass.getAnnotation(Timeout.class);
            OkHttpClient.Builder builder = getClient().newBuilder();
            if (timeout.connect() > 0) {
                builder.connectTimeout(timeout.connect(), timeout.timeUnit());
            }
            if (timeout.read() > 0) {
                builder.readTimeout(timeout.read(), timeout.timeUnit());
            }
            if (timeout.write() > 0) {
                builder.writeTimeout(timeout.write(), timeout.timeUnit());
            }
            retrofitBuilder.client(builder.build());
        }
    }

    @NonNull
    public <API> String getCurrentBaseUrl(@NonNull Class<API> api) {
        return baseUrls.get(api);
    }

    /**
     * get full url with append
     */
    public static <API> String getBaseUrl(@NonNull Class<API> api, @Nullable String newBaseUrl) {
        if (api.isAnnotationPresent(BaseUrl.class)) {
            BaseUrl baseUrl = api.getAnnotation(BaseUrl.class);
            String url = newBaseUrl != null ? newBaseUrl : baseUrl.value();
            if (!url.endsWith("/")) url += "/";

            String append = baseUrl.append();
            if (!TextUtils.isEmpty(append)) {
                if (append.startsWith("/")) append = append.substring(1);
                url = url + append;
            }
            if (!url.endsWith("/")) url += "/";
            return url;
        } else {
            throw new IllegalArgumentException(String.format("%s: MUST ANNOTATE WITH '%s'", api.getName(), BaseUrl.class.getName()));
        }
    }

    /**
     * eg: BaseUrl(value="https://www.a.com/", append = "apiA") => "https://www.a.com"
     *
     * @return current url without append & '/'
     */
    public <API> String getCurrentUrlWithoutAppend(@NonNull Class<API> api) {
        if (api.isAnnotationPresent(BaseUrl.class)) {
            BaseUrl baseUrl = api.getAnnotation(BaseUrl.class);
            String append = baseUrl.append();
            String url = getCurrentBaseUrl(api);
            if (url.endsWith("/")) url = url.substring(0, url.length() - 1);
            if (TextUtils.isEmpty(append)) {
                return url;
            } else {
                if (append.endsWith("/")) append = append.substring(0, append.length() - 1);
                url = url.substring(0, url.length() - append.length());
                if (url.endsWith("/")) url = url.substring(0, url.length() - 1);
                return url;
            }
        } else {
            throw new IllegalArgumentException(String.format("%s: MUST ANNOTATE WITH '%s'", api.getName(), BaseUrl.class.getName()));
        }
    }

    /**
     * get url from with @{@link BaseUrl}  with append
     */
    @NonNull
    public static <API> String getBaseUrl(@NonNull Class<API> api) {
        return getBaseUrl(api, null);
    }

    /**
     * hold all api in one
     */
    private <HOLDER> HOLDER createHolderApi(@NonNull final Map<Class, Object> apis, @NonNull Class<HOLDER> apiHolder) {
        return (HOLDER) Proxy.newProxyInstance(apiHolder.getClassLoader(), new Class[]{apiHolder}, new InvocationHandler() {
            @Override
            public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
                Object api = apis.get(method.getDeclaringClass());
                return invokeApi(api, method, args, isAndroidApi);
            }
        });
    }

    protected Object invokeApi(@NonNull Object api, @NonNull Method method, Object[] args, boolean isAndroidApi) throws Throwable {
        Object result = method.invoke(api, args);
        return isAndroidApi ? setAndroidSchedulers(result) : result;
    }

    protected Object setAndroidSchedulers(@NonNull Object result) {
        if (result instanceof Single) {
            return ((Single) result).compose(new SingleTransformer() {
                @Override
                public SingleSource apply(Single upstream) {
                    return upstream.subscribeOn(Schedulers.io())
                            .unsubscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread());
                }
            });
        } else if (result instanceof Observable) {
            return ((Observable) result).compose(new ObservableTransformer() {
                @Override
                public ObservableSource apply(Observable upstream) {
                    return upstream.subscribeOn(Schedulers.io())
                            .unsubscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread());
                }
            });
        }
        return result;
    }
}

File: holder/src/main/java/com/dhy/apiholder/ApiHolderUtil.kt
package com.dhy.apiholder

import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.lang.reflect.Method
import java.lang.reflect.Proxy
import java.net.URL
import kotlin.reflect.KClass

open class ApiHolderUtil<HOLDER : Any>(private val holder: KClass<HOLDER>, private val validateEagerly: Boolean = false) {
    val api: HOLDER by lazy { createHolderApi(holder) }
    private val okHttpClient by lazy { getClient() }
    private val retrofit by lazy { getRetrofit(okHttpClient) }
    private val apis: MutableMap<Class<*>, Any> = mutableMapOf()
    private val baseUrls: MutableMap<Class<*>, String> = mutableMapOf()
    protected open fun getClient(): OkHttpClient = OkHttpClient()
    protected open fun getGsonConverterFactory(): GsonConverterFactory = GsonConverterFactory.create()
    protected open fun getRetrofit(client: OkHttpClient): Retrofit {
        return Retrofit.Builder()
                .validateEagerly(validateEagerly)
                .addConverterFactory(getGsonConverterFactory())
                .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
                .baseUrl("http://www.demo.com/")
                .client(client)
                .build()
    }

    private fun createHolderApi(apiHolder: KClass<HOLDER>): HOLDER {
        val partApis = apiHolder.java.interfaces
        for (apiClass in partApis) updateApi(apiClass)
        onCreateHolderApi()
        return createHolderApi(apiHolder.java)
    }

    open fun onCreateHolderApi() {}

    /**
     * the domain end which is '/' will be ignored
     * */
    fun updateApi(releaseDomain: String, newDomain: String) {
        val release = releaseDomain.formatEnd()
        val news = newDomain.formatEnd()
        val apis = holder.java.interfaces.filter {
            val url = getAnnotationUrl(it).value.formatEnd()
            //DynamicServer may update 'release' any time, so check 'news' also.
            url == release || url == news
        }
        if (apis.isNotEmpty()) {
            apis.forEach {
                updateApi(it, newDomain)
            }
        }
    }

    fun updateApi(apiClass: KClass<*>, newDomain: String? = null) {
        updateApi(apiClass.java, newDomain)
    }

    private fun updateApi(apiClass: Class<*>, newDomain: String? = null) {
        val oldUrl = baseUrls[apiClass]
        val newUrl = getBaseUrl(apiClass, newDomain).toString()
        if (oldUrl != newUrl) {
            baseUrls[apiClass] = newUrl
            if (apis[apiClass] != null) {
                apis[apiClass] = createApi(apiClass)
            }
        }
    }

    private fun createApi(apiClass: Class<*>): Any {
        val url = baseUrls.getValue(apiClass)
        val retrofitBuilder = retrofit.newBuilder()
        retrofitBuilder.initTimeout(apiClass)
        val retrofit = retrofitBuilder.baseUrl(url).build()
        return retrofit.create(apiClass)
    }

    private fun Retrofit.Builder.initTimeout(apiClass: Class<*>) {
        if (apiClass.isAnnotationPresent(Timeout::class.java)) {
            val timeout = apiClass.getAnnotation(Timeout::class.java)!!
            val builder = okHttpClient.newBuilder()
            if (timeout.connect > 0) {
                builder.connectTimeout(timeout.connect, timeout.timeUnit)
            }
            if (timeout.read > 0) {
                builder.readTimeout(timeout.read, timeout.timeUnit)
            }
            if (timeout.write > 0) {
                builder.writeTimeout(timeout.write, timeout.timeUnit)
            }
            client(builder.build())
        }
    }

    private fun getSingleApi(apiClass: Class<*>): Any {
        val api = apis[apiClass]
        return if (api != null) api
        else {
            val newApi = createApi(apiClass)
            apis[apiClass] = newApi
            newApi
        }
    }

    /**
     * hold all api in one
     */
    @Suppress("UNCHECKED_CAST")
    private fun createHolderApi(apiHolder: Class<HOLDER>): HOLDER {
        return Proxy.newProxyInstance(apiHolder.classLoader, arrayOf<Class<*>>(apiHolder)) { _, method, args ->
            val api = getSingleApi(method.declaringClass)
            invokeApi(api, method, args)
        } as HOLDER
    }

    @Throws(Throwable::class)
    protected open fun invokeApi(api: Any, method: Method, args: Array<*>?): Any {
        return if (args != null) method.invoke(api, *args)
        else method.invoke(api)
    }

    /**
     * eg: BaseUrl(value="domain", append = "a") => "domain"
     *
     * @return using domain without append & '/'
     */
    fun getUsingDomain(api: KClass<*>): String {
        val cls = api.java
        val url = getBaseUrl(cls.kotlin, false).toString()
        val append = getAnnotationUrl(cls).append
        return url.replace("/?$append/?\$".toRegex(), "")
    }

    /**
     * @param release release or current using full url
     * */
    fun getBaseUrl(api: KClass<*>, release: Boolean): URL {
        return if (baseUrls.isEmpty() || release) getBaseUrl(api.java)
        else {
            val url = baseUrls[api.java]
            if (url != null) URL(url)
            else getBaseUrl(api.java)
        }
    }

    /**
     * get full url with append, end with "/"
     */

    fun getBaseUrl(api: KClass<*>, newDomain: String? = null): URL {
        return getBaseUrl(api.java, newDomain)
    }

    /**
     * get full url with append, end with "/"
     */

    fun getBaseUrl(api: Class<*>, newDomain: String? = null): URL {
        val baseUrl = getAnnotationUrl(api)
        val domain = newDomain ?: baseUrl.value
        return URL(domain.appendPath(baseUrl.append))
    }

    private val annotations: MutableMap<Class<*>, IBaseUrl> = mutableMapOf()
    private fun getAnnotationUrl(cls: Class<*>, annotationBuffer: MutableMap<Class<*>, IBaseUrl> = annotations): IBaseUrl {
        val buffer = annotationBuffer[cls]
        return if (buffer != null) buffer
        else {
            val baseUrl = if (cls.isAnnotationPresent(BaseUrl::class.java)) {
                val data = cls.getAnnotation(BaseUrl::class.java)!!
                BaseUrlData(data.value, data.append)
            } else {
                getUserBaseUrl(cls)
            }
            annotationBuffer[cls] = baseUrl
            baseUrl
        }
    }

    open fun getUserBaseUrl(cls: Class<*>): IBaseUrl {
        throw IllegalArgumentException("not supported yet")
    }
}

private class BaseUrlData(override val value: String, override val append: String) : IBaseUrl

fun String.trim(tail: String): String {
    return if (endsWith(tail)) {
        substring(0, length - tail.length)
    } else this
}

private fun String.formatEnd(): String {
    return appendPath()
}

/**
 * appendPath for URL
 * @return end with "/"
 * */
fun String.appendPath(path: String? = null): String {
    val url = this + "/" + (path ?: "") + "/"
    return url.replace("([^:/])/{2,}".toRegex(), "$1/")
}